### PR TITLE
Regard non-existing dependency-profile-id and return None, if no hand…

### PR DESCRIPTION
**Pull request by ida**
…led case applied, fixes https://github.com/plone/Products.CMFPlone/issues/2228.

Ideally non-existing dependencies should not be written into a profile's info-property, but I coudn't find the place where that happens.

### CODE REVIEW

File: Products/GenericSetup/tool.py [hunk 1]
```diff
@@ -921,17 +921,30 @@ def getProfileInfo(self, profile_id):
 
     @security.protected(ManagePortal)
     def getDependenciesForProfile(self, profile_id):
-        if profile_id is None:
-            return ()
-        if profile_id.startswith("snapshot-"):
-            return ()
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: I don't know why the early exit here is wrong:  without a `profile_id` we cannot look up dependencies, and snapshots never have dependencies (they are full exports of a site's state, not intended to be composed with others).

File: Products/GenericSetup/tool.py [hunk 2]
```diff
@@ -921,17 +921,30 @@ def getProfileInfo(self, profile_id):
...
 
-        if not self.profileExists(profile_id):
-            raise KeyError(profile_id)
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: I can't see why raising for a non-existent profile is wrong, either.

File: Products/GenericSetup/tool.py [hunk 3]
```diff
@@ -921,17 +921,30 @@ def getProfileInfo(self, profile_id):
...
-        try:
-            return self.getProfileInfo(profile_id).get('dependencies', ())
-        except KeyError:
-            return ()
+        dependencies = None
+        profile_info = None
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: These assignments are redundant.

File: Products/GenericSetup/tool.py [hunk 4]
```diff
@@ -921,17 +921,30 @@ def getProfileInfo(self, profile_id):
...
+
+        if profile_id is None or profile_id.startswith("snapshot-"):
+            dependencies = ()
+
+        profile_info = self.getProfileInfo(profile_id)
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: I'm assuming that you want the KeyError raised by this line to propagate:  is that correct?

File: Products/GenericSetup/tool.py [hunk 5]
```diff
@@ -921,17 +921,30 @@ def getProfileInfo(self, profile_id):
...
+
+        if 'dependencies' not in profile_info:
+            dependencies = ()
+        else:
+            dependencies = profile_info.get('dependencies')
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: `profile_info` is a dict:  just use `dependencies profile_info.get('dependencies', ()` rather than the `if` test.

File: Products/GenericSetup/tool.py [hunk 6]
```diff
@@ -921,17 +921,30 @@ def getProfileInfo(self, profile_id):
...
+
+            for dependency_id in dependencies:
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: Move this out to the top level (where it will be a no-op if `dependencies` is empty).

File: Products/GenericSetup/tool.py (hunk 1 above)
Comment (ida) [reply Products/GenericSetup/tool.py]: It's not wrong, just wanted to save two lines. But yes, saving processing time is better, thanks.

File: Products/GenericSetup/tool.py (hunk 2 above)
Comment (ida) [reply Products/GenericSetup/tool.py]: Not wrong, just superfluous, because instead in line 931 `getProfileInfo` will already fail with a KeyError, which also contains the profile-id, nothing more to do here, AFAICS.

File: Products/GenericSetup/tool.py (hunk 6 above)
Comment (tseaver) [reply Products/GenericSetup/tool.py]: You could just push updates to this PR's branch, which would actually be easier to re-review.

File: Products/GenericSetup/tool.py [hunk 7]
```diff
@@ -921,17 +921,30 @@ def getProfileInfo(self, profile_id):
 
     @security.protected(ManagePortal)
     def getDependenciesForProfile(self, profile_id):
+
+        dependencies = None
+        profile_info = None
+
         if profile_id is None:
-            return ()
+            dependencies = ()
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: I still prefer the early return here:  there is no point in calling `getProfileInfo` for either of these cases.

File: Products/GenericSetup/tool.py [hunk 8]
```diff
@@ -921,17 +921,30 @@ def getProfileInfo(self, profile_id):
...
+
         if profile_id.startswith("snapshot-"):
-            return ()
+            dependencies = ()
+
+        profile_info = self.getProfileInfo(profile_id)
+
+        dependencies = profile_info.get('dependencies', ())
+
+            for dependency_id in dependencies:
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: Unindent here (this is a `SyntaxError`).

Comment (ida) [reply Products/GenericSetup/tool.py]: My bad, used the web-editor, that produced tabs instead of spaces.
Making them spaces was not recognized as a changement, so added a space somewhere else, to make it commitable. Travis is super-fast, wow.

**Comment by ida** [comment]
Not sure when I can find time for that, is it ok to leave the PR like this for now?

**Comment by ida** [comment]
Sorry, thought it was green, but the next indent-err ocurred. It's kinda late here and it was an exhausting day, probably just should get some sleep, before continuing (and not use the web-editor).

Have a great day!

File: Products/GenericSetup/tool.py [hunk 9]
```diff
@@ -915,23 +915,52 @@ def profileExists(self, profile_id):
...
+
     @security.protected(ManagePortal)
     def getProfileInfo(self, profile_id):
         return _profile_registry.getProfileInfo(profile_id)
 
     @security.protected(ManagePortal)
     def getDependenciesForProfile(self, profile_id):
+
+        dependencies = None
+        profile_info = None
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: Both variables are assigned unconditionally below -- no need to initialize them here.

File: Products/GenericSetup/tool.py [hunk 10]
```diff
@@ -915,23 +915,52 @@ def profileExists(self, profile_id):
...
+            if not self.profileExists(profile_id):
+                return False
+        return True
+
+    security.declareProtected(ManagePortal, "dependenciesExist")
+    def dependenciesExist(self, profile_id):
+        """Check if all dependency-profiles exist."""
+        try:
+            dependency_ids = self.getDependenciesForProfile(profile_id)
+        except:
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: Avoid bare `except:`:  use `except KeyError:`.

File: Products/GenericSetup/tests/test_tool.py [hunk 11]
```diff
@@ -1706,21 +1714,26 @@ def test_getDependenciesForProfile(self):
...
+            'baz', 'Baz', '', self._PROFILE_PATH3, profile_type=EXTENSION)
 
         self.assertEqual(tool.getDependenciesForProfile('other:foo'),
                          (u'profile-other:bar', ))
         self.assertEqual(tool.getDependenciesForProfile('other:bar'), ())
         self.assertEqual(tool.getDependenciesForProfile('snapshot-some'), ())
         self.assertEqual(tool.getDependenciesForProfile(None), ())
         self.assertRaises(KeyError, tool.getDependenciesForProfile, 'nonesuch')
+        # profile_info does contain dependencies, but one of them doesn't exist:
+        self.assertEqual(tool.dependenciesExist('other:baz'), False)
```
Comment (tseaver) [inline Products/GenericSetup/tests/test_tool.py]: This bit really belongs in a separate testcase.  Given that it is here:

- Use `self.assertFalse`.
- Add assertions for the other two profiles.

E.g.:

```python
        # profile_info contains dependencies, and they exist.
        self.assertTrue(tool.dependenciesExist('other:foo'))
        # profile does not contain dependencies
        self.assertTrue(tool.dependenciesExist('other:bar'))
        # profile_info contains dependencies, but one of them doesn't exist:
        self.assertFalse(tool.dependenciesExist('other:baz'))
```

File: Products/GenericSetup/tool.py (hunk 9 above)
Comment (ida) [reply Products/GenericSetup/tool.py]: Oh yeah, thanks. Think I got overcautious, because it was the cause in the bug of CMFPlone.

File: Products/GenericSetup/tests/test_tool.py (hunk 11 above)
Comment (ida) [reply Products/GenericSetup/tests/test_tool.py]: Agree. Thanks for the corrections! Gonna be on it tomorrow again.

File: Products/GenericSetup/tool.py (hunk 10 above)
Comment (ida) [reply Products/GenericSetup/tool.py]: Oops, jupp.

File: Products/GenericSetup/tool.py [hunk 12]
```diff
@@ -939,16 +942,36 @@ def getDependenciesForProfile(self, profile_id):
...
+        dependency_ids = self.getDependenciesForProfile(profile_id,
+                                                ignore_broken=True)
+        for dependency_id in dependency_ids:
+            if not self.profileExists(dependency_id):
+                broken_dependencies.append(dependency_id)
+        return broken_dependencies
+
+    @security.protected(ManagePortal)
+    def hasDependencies(self, profile_id):
+        return len(self.getDependenciesForProfile(profile_id)) > 0
```
Comment (tseaver) [inline Products/GenericSetup/tool.py]: I see you've added a test for `hasBrokenDependencies` (and therefore cover `getBrokenDependencies`, but does `hasDependencies` need to exist?  If so, it needs tests (empty deps, normal deps, one or more broken deps).

Comment (ida) [reply Products/GenericSetup/tool.py]: Oh gee, that's a leftover not used anywhere, I remove it.

**Comment by ida** [comment]
@tseaver A pleasure. With such good advice it's real fun to fix bugs.

**Comment by icemac** [comment]
Should this fix be mentioned in the change log, too?

**Comment by ida** [comment]
@icemac Of course :-) Adding it,,, Thanks!
